scripts: fix release note script for bors

couchand opened the pull request:
Since Bors merges with a slightly different message and a
significantly different strategy compared to GitHub, the
pattern matching and logic for the release notes script
was broken by the introduction of the Bors workflow.  This
fixes the script to account for the various merge types.

Merged PRs are now found by git ref.  This requires that
you configure your local repo to pull the PR refs from
GitHub.  To do this, add a line like:

    fetch = +refs/pull/*/head:refs/pull/origin/*

to the GitHub remote section of .git/config, then run
`git fetch -a` to pull them down.

Fixes: #24418
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24532)
<!-- Reviewable:end -->

knz commented:
ping me when you want me to review this. When that time comes, remove the old code first.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-:-L9MltCi1qT_4nm4qREY:bjsbder)*
<!-- Sent from Reviewable.io -->

couchand committed:
scripts: clean up formatting of contributors section

Previously, the formatting of the Contributors section of the
generated release notes was off in the case that no external
contributors authored commits for a given release.  This fixes
the formatting in that case.

Release note: None

couchand committed:
scripts: fix release notes for Bors

Since Bors merges with a slightly different message and a
significantly different strategy compared to GitHub, the
pattern matching and logic for the release notes script
was broken by the introduction of the Bors workflow.  This
fixes the script to account for the various merge types.

Merged PRs are now found by git ref.  This requires that
you configure your local repo to pull the PR refs from
GitHub.  To do this, add a line like:

    fetch = +refs/pull/*/head:refs/pull/origin/*

to the GitHub remote section of .git/config, then run
`git fetch -a` to pull them down.

Fixes: #24418
Release note: None

couchand committed:
scripts: add some new roachers to release notes

This adds Bob, Lauren, and Victor to the roachers list, as well as
adding an alias for Amruta.

Release note: None

couchand committed:
scripts: validate release note options

This adds a bit of validation to the release note script
options, to print more useful errors if something is wrong.

Release note: None

couchand committed:
scripts: release notes handle mid-PR branch merges

Previously, the release notes script would just give up on
mid-PR branch merges.  This change fixes that, to make sure
to include changes on a PR only reachable from that PR.

Release note: None

couchand commented:
Ok @knz this is ready to review.

There was one situation I noticed during testing that I noted with a TODO but didn't actually try to solve here.  PR #19473 doesn't have a merge commit -- it looks like it was squash-merged through the GitHub web interface.  It shouldn't be too hard to extend the script to handle that, but since I didn't see an example since October, I didn't spend much time on it.

couchand committed:
scripts: remove irrelevant comment

When testing the updates to the release note script, I noticed
that squash-merges weren't handled correctly, but I didn't see
one since last October.  As it turns out, that option has been
disabled, so we shouldn't be seeing any more of them.

Release note: None

knz commented:
The code looks good modulo a perhaps nit and a question, see below.

Please run the end solution by Jesse and/or Amruta to check whether the change works well for them (I expect there will be no surprise). No real need to show them the per-author list as they usually don't use it. I am happy with that result for now.

---

Reviewed 1 of 1 files at r10.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[scripts/release-notes.py, line 14 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eflA3BPsQDSNnczqf:-L9eflA3BPsQDSNnczqg:b1dhiq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L14)):*
> ```Python
> #     GitHub.  To do this, add a line like:
> #       fetch = +refs/pull/*/head:refs/pull/origin/*
> #     to the GitHub remote section of .git/config.
> ```

Please copy this change in the commit message, and give a small suggestion to the doc team in the commit message about what they need to do _assuming they already have a clone_: 1) change the git config 2) git fetch -a

---

*[scripts/release-notes.py, line 75 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eg0wi-i0mtFgdbi-4:-L9eg0wi-i0mtFgdbi-5:byqc56a) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L75)):*
> ```Python
>     "Arjun Narayan",
>     "Ben Darnell",
>     "Bob Vawter",
> ```

omg yes sorry we had forgotten that.

---

*[scripts/release-notes.py, line 279 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eiSQkAPaCmFLqjDRP:-L9eiSQkAPaCmFLqjDRQ:b13s2do) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L279)):*
> ```Python
>     item = {'author': (commit.author.name, commit.author.email),
>             'sha': commit.hexsha[:shamin],
>             'pr': "#{0}".format(pr),
> ```

here and below: 
1) `pr` is a string so simply do `'#' + pr`. No need to over-engineer.
2) since `pr` is parsed with `#` I do not understand why you felt the need to strip the `#` upfront. It seems to me you could keep it throughout and only strip it when constructing the pull tip label.

---

*[scripts/release-notes.py, line 374 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9ehThZ-7qz026JC-8C:-L9ehThZ-7qz026JC-8D:b-ydlcik) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L374)):*
> ```Python
> 
>         # Exclude any parents reachable from the other side of the
>         # PR merge commit.
> ```

Can you briefly explain to me here (maybe as a review comment) what this does precisely.

Also explain to me what happens if an engineer has secondary branches and merges from their secondary branch into their main feature branch (non-master) before they send the PR. Are both sides correctly explored in that case?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-:-L9eikXc8lkpsxnTRf3K:bf81qyt)*
<!-- Sent from Reviewable.io -->

couchand committed:
script: add explanatory comment to git-fu in release notes

Release note: None

couchand commented:
@knz not sure if you're asking me to run anything specific by Jesse & Amruta.  I will alert them to this work, but in the end there should be no visible changes except to the per-author list.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[scripts/release-notes.py, line 14 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eflA3BPsQDSNnczqf:-L9eqasK5qXNjaTNeZcd:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L14)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please copy this change in the commit message, and give a small suggestion to the doc team in the commit message about what they need to do _assuming they already have a clone_: 1) change the git config 2) git fetch -a
</blockquote></details>

done.

---

*[scripts/release-notes.py, line 75 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eg0wi-i0mtFgdbi-4:-L9eqd347ANjDFlUlnXv:b2ni8oq) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L75)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

omg yes sorry we had forgotten that.
</blockquote></details>

np!

---

*[scripts/release-notes.py, line 279 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9eiSQkAPaCmFLqjDRP:-L9eqdrt7ux8cPclJ4bW:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L279)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

here and below: 
1) `pr` is a string so simply do `'#' + pr`. No need to over-engineer.
2) since `pr` is parsed with `#` I do not understand why you felt the need to strip the `#` upfront. It seems to me you could keep it throughout and only strip it when constructing the pull tip label.
</blockquote></details>

done

---

*[scripts/release-notes.py, line 374 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9ehThZ-7qz026JC-8C:-L9eqewEE4ZeYH5Z1piN:b-2efb6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L374)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you briefly explain to me here (maybe as a review comment) what this does precisely.

Also explain to me what happens if an engineer has secondary branches and merges from their secondary branch into their main feature branch (non-master) before they send the PR. Are both sides correctly explored in that case?
</blockquote></details>

Yes, this is definitely hairy and may have bugs.  It works on our recent repository history as far as I can tell, but we may need to revisit it in the future if other weird merge situations come up.

I've now added this explanation as a code comment.

I believe it should correctly visit all commits that are on the PR and only on the PR.  If there's some secondary branch merge included on the PR, if those commits don't otherwise end up reachable from the target branch, they'll be included.  If there's a back-merge from the target branch, that should be excluded.  Examples:

### secondary branch merged into PR

Dev branched off of K, made a commit J, made a commit G while someone else committed H, merged H from the secondary branch to the topic branch in E, made a final commit in C, then merged to master in A.

```
A <-- master
|\
| \
B  C <-- PR tip
|  |
|  |
D  E <-- secondary merge
|  |\
|  | \
F  G  H <-- secondary branch
|  | /
|  |/
I  J
| /
|/
K <-- merge base
```

C, E, G, H, and J will each be checked.  None of them are reachable from B, so they will all be visited.  E will be not be counted because the message starts with "Merge", so in the end C, G, H, and J will be included.

### back-merge from target branch

Dev branched off F, made one commit G, merged the latest from master in E, made one final commit in C, then merged the PR.

```
A <-- master
|\
| \
B  C <-- PR tip
|  |
|  |
D  E <-- back-merge
| /|
|/ |
F  G
| /
|/
H <-- merge base
```

C, E, F, and G will each be checked.  F is reachable from B, so it will be excluded.  E starts with "Merge", so it will not be counted.  Only C and G will have statistics included.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-:-L9evtk2-NVSTullqayo:b-au4l7d)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but please consider a test as suggested below.

---

Reviewed 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[scripts/release-notes.py, line 374 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-L9ehThZ-7qz026JC-8C:-L9ex1dgCmsoceyXH-Dc:b-lqkkfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/01cf318fd921e366d879aef9cf94e29301f7b673/scripts/release-notes.py#L374)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Yes, this is definitely hairy and may have bugs.  It works on our recent repository history as far as I can tell, but we may need to revisit it in the future if other weird merge situations come up.

I've now added this explanation as a code comment.

I believe it should correctly visit all commits that are on the PR and only on the PR.  If there's some secondary branch merge included on the PR, if those commits don't otherwise end up reachable from the target branch, they'll be included.  If there's a back-merge from the target branch, that should be excluded.  Examples:

### secondary branch merged into PR

Dev branched off of K, made a commit J, made a commit G while someone else committed H, merged H from the secondary branch to the topic branch in E, made a final commit in C, then merged to master in A.

```
A <-- master
|\
| \
B  C <-- PR tip
|  |
|  |
D  E <-- secondary merge
|  |\
|  | \
F  G  H <-- secondary branch
|  | /
|  |/
I  J
| /
|/
K <-- merge base
```

C, E, G, H, and J will each be checked.  None of them are reachable from B, so they will all be visited.  E will be not be counted because the message starts with "Merge", so in the end C, G, H, and J will be included.

### back-merge from target branch

Dev branched off F, made one commit G, merged the latest from master in E, made one final commit in C, then merged the PR.

```
A <-- master
|\
| \
B  C <-- PR tip
|  |
|  |
D  E <-- back-merge
| /|
|/ |
F  G
| /
|/
H <-- merge base
```

C, E, F, and G will each be checked.  F is reachable from B, so it will be excluded.  E starts with "Merge", so it will not be counted.  Only C and G will have statistics included.
</blockquote></details>

Yes this description / spec is good and I agree with it.
I'm just not sure how the code is achieving this yet :)

TBH that's the point I'd suggest a mini-test that validates the behavior. You can perhaps use zip files embedding test repos.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-:-L9exJPP3999zu1pWt1E:bymrhmg)*
<!-- Sent from Reviewable.io -->

couchand commented:
That sounds fun.  I suggest we do that when some history comes up that this doesn't account for :trollface: 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24532#-:-L9exrzS1trpPi4hlrRL:b-wt6xwf)*
<!-- Sent from Reviewable.io -->

couchand commented:
bors r+

couchand commented:
bors ping

couchand commented:
> But how do you reckon we will even recognize when this situation occurs?

¯\_(ツ)_/¯

Does anybody look through the release notes to check them for accuracy?  Do we just assume the script works fine?

knz commented:
No the accuracy is not checked. That's why I was concerned.

couchand commented:
> No the accuracy is not checked. That's why I was concerned.

well, I can write a test for the testing i already did, which would prevent us from breaking it, but I don't think that's the main concern here (though it is a valid concern...).   The bigger issue is that we likely won't catch new issues that come up that I'm not predicting, which would be hard to write a test for...

couchand commented:
shouldn't have played with fire restarting bors while this was building...

bors r+

couchand commented:
bors ping

couchand commented:
bors ping

knz commented:
I think you can assuage my fears readily as follows:

- ask the git repo about the total number of commits encountered between the start and end points
- while you're exploring the PR branches, count the number of commits effectively analyzed and accounted to each PR
- at the end, verify that the total number of commits in the repo = the number of commits you've assigned to each PR

and report any discrepancy as a list of commits (possibly including their commit message) that need further attention. This is actionable even without a fix in the code, as the doc team will then be able to extract documentable information from these "orphan" commits.

couchand commented:
created an issue to follow up on that.

bors r+

knz commented:
:+1: thank you!
